pallets/amm/src/lib.rs

```diff
@@ -103,21 +119,149 @@ mod pallet {
+
+            let amm_pool = PoolLiquidityAmount {
+                base_amount,
+                quote_amount,
+            };
+            Pools::<T, I>::insert(base_asset, &quote_asset, amm_pool.clone());
+            LiquidityProviders::<T, I>::insert((who.clone(), base_asset, quote_asset), amm_pool);
+
+            T::Currency::transfer(base_asset, &who, &Self::account_id(), base_amount)?;
+            T::Currency::transfer(quote_asset, &who, &Self::account_id(), quote_amount)?;
+
+            Self::deposit_event(Event::<T, I>::LiquidityAdded(who, base_asset, quote_asset));
+            Ok(().into())
+        }
+
+        /// Allow users to remove liquidity from a given pool
+        ///
+        /// - `pool`: Currency pool, in which liquidity will be removed
+        /// - `liquidity_amounts`: Liquidity amounts to be removed from pool
+        #[pallet::weight(10_000)]
+        pub fn remove_liquidity(
+            origin: OriginFor<T>,
+            pool: (CurrencyId, CurrencyId),
+            liquidity_amounts: (Balance, Balance),
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            let sorted_assets = &Self::get_upper_currency(pool.0, pool.1);
+
+            let base_asset = sorted_assets.0;
+            let quote_asset = sorted_assets.1;
+            let base_amount = liquidity_amounts.0;
+            let quote_amount = liquidity_amounts.1;
+
+            Pools::<T, I>::try_mutate(
+                &base_asset,
+                &quote_asset,
+                |pool_liquidity_amount| -> DispatchResult {
+                    pool_liquidity_amount.base_amount = pool_liquidity_amount
+                        .base_amount
+                        .checked_sub(base_amount)
+                        .ok_or(ArithmeticError::Underflow)?;
+                    pool_liquidity_amount.quote_amount = pool_liquidity_amount
+                        .quote_amount
+                        .checked_sub(quote_amount)
+                        .ok_or(ArithmeticError::Underflow)?;
+                    Ok(())
+                },
+            )?;
+
+            LiquidityProviders::<T, I>::try_mutate(
+                (who.clone(), base_asset, quote_asset),
+                |pool_liquidity_amount| -> DispatchResult {
+                    pool_liquidity_amount.base_amount = pool_liquidity_amount
+                        .base_amount
+                        .checked_sub(base_amount)
+                        .ok_or(ArithmeticError::Underflow)?;
+                    pool_liquidity_amount.quote_amount = pool_liquidity_amount
+                        .quote_amount
+                        .checked_sub(quote_amount)
+                        .ok_or(ArithmeticError::Underflow)?;
+                    Ok(())
+                },
+            )?;
+
+            T::Currency::transfer(base_asset, &Self::account_id(), &who, base_amount)?;
+            T::Currency::transfer(quote_asset, &Self::account_id(), &who, quote_amount)?;
```

Comment:
We actually don't withdraw `base_amount` and `quote_amount` from the pool. The amount sent back should be proportional to the amount of ownership of the pool by `origin`

Reply:
@ETeissonniere Could you please suggest how should I calculate this?

Reply:
https://www.notion.so/parallelfinance/Ownership-and-Liquidity-9ae44353a5f24fe6bd02640cf41ed0af

Reply:
https://www.notion.so/parallelfinance/Ownership-and-Liquidity-9ae44353a5f24fe6bd02640cf41ed0af